python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -80,10 +88,20 @@ def dispatch_next_task(self) -> None:
         """Move a bundle from the operator inqueue to the operator itself."""
         for i, inqueue in enumerate(self.inqueues):
             if inqueue:
-                self.op.add_input(inqueue.pop(0), input_index=i)
+                self.op.add_input(inqueue.popleft(), input_index=i)
                 return
         assert False, "Nothing to dispatch"
 
+    def get_output_blocking(self) -> MaybeRefBundle:
+        """Get an item from this node's output queue, blocking as needed.
+
+        Returns:
+            The RefBundle from the output queue, or an error / end of stream indicator.
+        """
+        while not self.outqueue:
+            time.sleep(0.01)
+        return self.outqueue.popleft()
```

Comment:
The CPython implementation of `deque` is thread-safe for most ops, and this LBYL should technically work since we only have a single consumer thread, but I'd still vote for switching to EAFP to make this a bit more defensive:
```suggestion
        while True:
            try:
                return self.outqueue.popleft()
            except IndexError:
                time.sleep(0.01)
```

Reply:
Also, did this sleep-and-pop loop on the `deque` end up being faster than having an intermediate synchronized `Queue` that we do a blocking `.get()` on? I'd expect Python-level busy-waiting with a 10ms sleep to be slower than `Queue`'s waiting for the underlying "not empty" condvar, unless the queue hopping is a surprising amount of overhead.

Reply:
The performance gain turned out to be avoiding that 100ms stall from the ray.wait(). In the previous design, the executor thread would only be able to push 1 item to the consumer thread in each iteration since the intra-thread queue had size 1. The new design allows the consumer to take items as fast as it can.

Reply:
Ah that makes more sense! I actually wonder if keeping `outqueue` as a `deque` and using the synchronized `Queue` as a thread-safe proxy would end up being faster than this Python-level busy-waiting... No need to try it out since I'm guessing that it won't be a bottleneck, but I think I'll try that out locally since I'm curious.